### PR TITLE
Prevent DefaultPassageFormatter from taking shorter overlapping passages

### PR DESCRIPTION
**Pull request by zkendall**
### Description

There is a bug in the DefaultPassageFormatter. It will take an passage even if it results in a shorter or more fragmented passage (highlight). The one scenario I have is when there is a phrase match which overlaps with some term matches.

This is demonstrated in the attached unit tests.

### CODE REVIEW

File: lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestDefaultPassageFormatter.java [hunk 1]
```diff
@@ -52,4 +53,25 @@ public void testBasic() throws Exception {
             + "&lt;&#x2F;div&gt; of this <u>very</u> formatter.\u2026 It&#x27;s not <u>very</u> N&#x2F;A!",
         formatter.format(passages, text));
   }
+
```
Comment (zkendall) [inline lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestDefaultPassageFormatter.java:56]: Added new test. Essentially a copy scenario of what is tested in the other modified test (TestUnifiedHighlighterStrictPhrases). However, this tests directly the changed class (DefaultPassageFormatter), so I figured worth the addition.

File: lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighterStrictPhrases.java [hunk 2]
```diff
@@ -183,13 +183,15 @@ public void testBasics() throws IOException {
   }
 
   public void testWithSameTermQuery() throws IOException {
```
Comment (zkendall) [inline lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighterStrictPhrases.java:185]: Edit existing test. The test was already testing having overlapping term and phrase queries; however it didn't reproduce the bug until I expanded the overlaps with multiple terms.

Without the fix, TestUnifiedHighlighterStrictPhrases fails like this
```
arrays first differed at element [0]; expected:<<b>Yin yang[ ]loooooooooong</b>, <...> but was:<<b>Yin yang[</b> <b>]loooooooooong</b>, <...>
```

File: lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighter.java [hunk 3]
```diff
@@ -1422,7 +1422,7 @@ public Analyzer getWrappedAnalyzer(String fieldName) {
             Set.of("field_tripples", "field_characters"),
             "danc",
             "<b>danc</b>e with star",
-            "<b>da</b><b>n</b><b>c</b>e with star");
+            "<b>dan</b><b>c</b>e with star");
```
Comment (zkendall) [inline lucene/highlighter/src/test/org/apache/lucene/search/uhighlight/TestUnifiedHighlighter.java:1425]: This test is of "every character" is a token AND "every _three_ characters is a token".. So it's clear that "dan" is a better highlight than "da" and "n" separate.

File: lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/DefaultPassageFormatter.java [hunk 4]
```diff
@@ -76,10 +76,15 @@ public String format(Passage[] passages, String content) {
...
+        // It's possible to have overlapping terms.
+        //   Look ahead to expand 'end' past all overlapping.
+        //   Only take new end if it is larger than current end.
         while (i + 1 < passage.getNumMatches() && passage.getMatchStarts()[i + 1] < end) {
-          end = passage.getMatchEnds()[++i];
+          if (passage.getMatchEnds()[i + 1] > end) {
+            end = passage.getMatchEnds()[++i];
+          } else {
+            i++;
+          }
```
[the remark this answers is not in the document]
Comment (zkendall) [reply lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/DefaultPassageFormatter.java]: Yes! Makes sense and tests pass. Pushed this change. Thanks.

File: lucene/CHANGES.txt [hunk 5]
```diff
@@ -169,6 +169,8 @@ Bug Fixes
 * GITHUB#12878: Fix the declared Exceptions of Expression#evaluate() to match those
   of DoubleValues#doubleValue(). (Uwe Schindler)
 
+* GITHUB#13384: Fix highlighter to use longer passages instead of shorter individual terms. (Zack Kendall)
```
Comment (stefanvodita) [inline lucene/CHANGES.txt:172]: I think it's fine to move this down to 9.12 if it's a bug fix.

Comment (zkendall) [reply lucene/CHANGES.txt]: Ah, my branch was way behind heh. And I may have misunderstood how that file was organized when I did it. Done!

**Review by stefanvodita — APPROVED** [review]
Thank you for this contribution @zkendall, really appreciate the tests and additional explanations you provided. I'm not the most knowledgeable in this area of the code, but the change looks good as far as I can tell.

**Comment by zkendall** [comment]
@stefanvodita Thanks. I had to resolve a conflict after your approval. I see your approval wasn't dismissed, so I guess you don't have "write access" to approve. Do you know who does or how I can get their attention?

**Comment by stefanvodita** [comment]
@zkendall - I just wanted to wait a couple days to give anyone else who wanted to review a chance to do so. I've merged the change now.

File: lucene/CHANGES.txt [hunk 6]
```diff
@@ -280,6 +280,8 @@ Optimizations
 Bug Fixes
 ---------------------
 
+* GITHUB#13384: Fix highlighter to use longer passages instead of shorter individual terms. (Zack Kendall)
```
Comment (dsmiley) [inline lucene/CHANGES.txt:283]: Next time please indicate *which* highlighter; we have 3-4!

**Pull request merged**